homeassistant/components/media_player/firetv.py

```diff
@@ -195,60 +155,134 @@ def source_list(self):
         """Return a list of running apps."""
         return self._running_apps
 
+    @adb_wrapper
     def update(self):
         """Get the latest date and update device state."""
-        self._state = {
-            'idle': STATE_IDLE,
-            'off': STATE_OFF,
-            'play': STATE_PLAYING,
-            'pause': STATE_PAUSED,
-            'standby': STATE_STANDBY,
-            'disconnected': STATE_UNKNOWN,
-        }.get(self._firetv.state, STATE_UNKNOWN)
-
-        if self._state not in [STATE_OFF, STATE_UNKNOWN]:
-            self._running_apps = self._firetv.running_apps
-            self._current_app = self._firetv.current_app
-        else:
-            self._running_apps = None
-            self._current_app = None
-
+        try:
+            # Check if device is disconnected.
```

Comment:
This try-catch block is really huge. Could you split it up to more malleable pieces? That will also simplify the exception handling / troubleshooting in the future.